[REST API] Analytics for application passwords web authorization

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #8559, a continuation for #8618
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This adds the analytics for events related to webview-based REST API login.

Conversation in p1679454856318259-slack-C046HDZL87J

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

First, please open #8618 and have a look at the three test cases.

1. For the **"Success path"** case, please make sure the following are tracked: `application_passwords_authorization_web_view_shown`, `application_passwords_authorization_approved`
2. For the **"Failure path 1 : Application passwords authentication rejected"** case, please make sure the following are tracked: `application_passwords_authorization_web_view_shown`, `application_passwords_authorization_rejected`
3. For the **Failure path 2: Application passwords disabled** case, please make sure the following is tracked: `application_passwords_authorization_url_not_available`

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hafizrahman committed:
Add: tracking keys related to REST API via webview- logins.

hafizrahman committed:
Add: tracking for webview shown.

hafizrahman committed:
Add: tracking for application password connection rejected.

hafizrahman committed:
Add: tracking for application password connection accepted.

hafizrahman committed:
Add: tracking for application password auth URL not available.

hafizrahman committed:
Move tracking for application password auth URL not available to a better spot.

hafizrahman committed:
Remove unneeded event.

This is already tracked in a different way, see https://github.com/woocommerce/woocommerce-android/pull/8618/files?diff=split&w=0#diff-6041eda7bbcd79b8879ec4c62d7dcda35fcd19e039a58ae32a9792fb20cca5aeR292-R296

hafizrahman committed:
Merge branch 'issue/8599-application-passwords-web-auhorization' into issue/8599-tracks-for-application-passwords-web-authorization

hafizrahman committed:
Improve the way we track webview shown.

Previously, since it's inside a combine, the tracking can be done multiple times just for a single webview shown event.

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt:
```diff
@@ -164,9 +165,13 @@ class LoginSiteCredentialsViewModel @Inject constructor(
                 if (!isSuccess) {
                     fetchedSiteId.value = -1
                     state.value = State.NativeLogin
+
+                    analyticsTracker.track(AnalyticsEvent.APPLICATION_PASSWORDS_AUTHORIZATION_REJECTED)
                     triggerEvent(ShowSnackbar(R.string.login_site_credentials_web_authorization_connection_rejected))
                     return@launch
                 }
+
+                analyticsTracker.track(AnalyticsEvent.APPLICATION_PASSWORDS_AUTHORIZATION_APPROVED)
```
These events are also fired twice`application_passwords_authorization_approved` and `application_passwords_authorization_rejected` fired twice. You can see that in the logcat. That is because this function `onWebAuthorizationUrlLoaded` is invoked twice from inside `WCWebView` composable. There are 2 callbacks that trigger this `onWebAuthorizationUrlLoaded`:
```kotlin
                        override fun doUpdateVisitedHistory(view: WebView?, url: String?, isReload: Boolean) {
                            url?.let { onUrlLoaded(it) }
                        }

                        override fun onLoadResource(view: WebView?, url: String?) {
                            url?.let { onUrlLoaded(it) }
                        }
```
From my tests this function being called twice only impacts the tracking, and it looks like the feature works fine. So I won't block the PR but we need to look into this and fix it IMHO. Let me know what you think.

hafizrahman replied:
Thanks! Added a fix in 1d090b7860b843aa170f68c0b1243da9938b7016 and d25a1bedce39da0225b1905dfd0e462393674598

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsScreen.kt:
```diff
@@ -252,6 +256,7 @@ private fun WebAuthorizationScreen(
         }
 
         viewState.authorizationUrl != null -> {
+            onWebViewShown(viewState.authorizationUrl)
```
FYI, just in case this is unexpected. In the success path where we open the webview and click on approve app passwords the event `application_passwords_authorization_web_view_shown` is tracked twice. 

```
Tracked: application_passwords_authorization_web_view_shown, Properties: {"auth_url":"https:\/\/spurred-tapir.jurassic.ninja\/wp-admin\/authorize-application.php?app_name=com.woocommerce.android.dev.app-client.Google-Pixel-2-XL&success_url=woocommerce:\/\/login","blog_id":0,"is_wpcom_store":false,"is_debug":true}

AND 

Tracked: application_passwords_authorization_web_view_shown, Properties: {"auth_url":"https:\/\/spurred-tapir.jurassic.ninja\/wp-admin\/authorize-application.php?app_name=com.woocommerce.android.dev.app-client.Google-Pixel-2-XL&success_url=woocommerce:\/\/login","blog_id":0,"is_wpcom_store":false,"is_debug":true}

```

hafizrahman replied:
Added fix in 932c9a49f4e32d07d822b0c4bb6955e2b2bc6456

JorgeMucientes commented:
Hey @hafizrahman great job. I left a couple comments. One of them is a bit concerning and should be fixed, although depending on the urgency to merge the webview login feature, we could fix this elsewhere after merging. Let me know what you think. I won't block the feature for now.

hafizrahman committed:
Fix: Move tracking for "webview shown" to be triggered by "Try again with wp-admin" click.

This ensures that it is only ever called once per view.

hafizrahman committed:
Fix: Remove onWebViewShown() usage for tracking, as it's no longer needed.

hafizrahman committed:
Add: Option onPageFinished parameter to WCWebView.

This might seem close in usage with onLoadResource(), but the main purpose here is to have something that calls the function parameter only once, unlike `onUrlLoaded` that gets called twice by `onLoadResource` and `doUpdateVisitedHistory`.

See https://github.com/woocommerce/woocommerce-android/pull/8660/files#r1151742457 for more context.

hafizrahman committed:
Add: Make use of `onPageFinished` to ensure `onWebAuthorizationUrlLoaded` is not called twice each time.

This ensures that tracking inside it only gets called once.

JorgeMucientes commented:
Thanks for fixing the duplicated tracking @hafizrahman! I've retested all the feature and it all works as expected. As discussed over Slack i think the approach followed to fix the duplicated calls is the safest as it es well scoped to the feature changes. Well done. 

I'll let you deal with the merging of both PRs :shipit:
